// recharts/lib/state/cartesianAxisSlice.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.removeZAxis = exports.removeYAxis = exports.removeXAxis = exports.cartesianAxisReducer = exports.addZAxis = exports.addYAxis = exports.addXAxis = void 0;
var _toolkit = require("@reduxjs/toolkit");
var _immer = require("immer");
/**
 * Properties shared in X, Y, and Z axes
 */

/**
 * These are the external props, visible for users as they set them using our public API.
 * There is all sorts of internal computed things based on these, but they will come through selectors.
 *
 * Properties shared between X and Y axes
 */

/**
 * Z axis is special because it's never displayed. It controls the size of Scatter dots,
 * but it never displays ticks anywhere.
 */

var initialState = {
  xAxis: {},
  yAxis: {},
  zAxis: {}
};

/**
 * This is the slice where each individual Axis element pushes its own configuration.
 * Prefer to use this one instead of axisSlice.
 */
var cartesianAxisSlice = (0, _toolkit.createSlice)({
  name: 'cartesianAxis',
  initialState,
  reducers: {
    addXAxis(state, action) {
      state.xAxis[action.payload.id] = (0, _immer.castDraft)(action.payload);
    },
    removeXAxis(state, action) {
      delete state.xAxis[action.payload.id];
    },
    addYAxis(state, action) {
      state.yAxis[action.payload.id] = (0, _immer.castDraft)(action.payload);
    },
    removeYAxis(state, action) {
      delete state.yAxis[action.payload.id];
    },
    addZAxis(state, action) {
      state.zAxis[action.payload.id] = (0, _immer.castDraft)(action.payload);
    },
    removeZAxis(state, action) {
      delete state.zAxis[action.payload.id];
    }
  }
});
var {
  addXAxis,
  removeXAxis,
  addYAxis,
  removeYAxis,
  addZAxis,
  removeZAxis
} = cartesianAxisSlice.actions;
exports.removeZAxis = removeZAxis;
exports.addZAxis = addZAxis;
exports.removeYAxis = removeYAxis;
exports.addYAxis = addYAxis;
exports.removeXAxis = removeXAxis;
exports.addXAxis = addXAxis;
var cartesianAxisReducer = exports.cartesianAxisReducer = cartesianAxisSlice.reducer;
