// recharts/lib/cartesian/GraphicalItemClipPath.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.GraphicalItemClipPath = GraphicalItemClipPath;
exports.useNeedsClip = useNeedsClip;
var _react = _interopRequireDefault(require("react"));
var _chartLayoutContext = require("../context/chartLayoutContext");
var _hooks = require("../state/hooks");
var _axisSelectors = require("../state/selectors/axisSelectors");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function useNeedsClip(xAxisId, yAxisId) {
  var xAxis = (0, _hooks.useAppSelector)(state => (0, _axisSelectors.selectXAxisSettings)(state, xAxisId));
  var yAxis = (0, _hooks.useAppSelector)(state => (0, _axisSelectors.selectYAxisSettings)(state, yAxisId));
  var needClipX = xAxis && xAxis.allowDataOverflow;
  var needClipY = yAxis && yAxis.allowDataOverflow;
  var needClip = needClipX || needClipY;
  return {
    needClip,
    needClipX,
    needClipY
  };
}
function GraphicalItemClipPath(_ref) {
  var {
    xAxisId,
    yAxisId,
    clipPathId
  } = _ref;
  var offset = (0, _chartLayoutContext.useOffset)();
  var {
    needClipX,
    needClipY,
    needClip
  } = useNeedsClip(xAxisId, yAxisId);
  if (!needClip) {
    return null;
  }
  var {
    left,
    top,
    width,
    height
  } = offset;
  return /*#__PURE__*/_react.default.createElement("clipPath", {
    id: "clipPath-".concat(clipPathId)
  }, /*#__PURE__*/_react.default.createElement("rect", {
    x: needClipX ? left : left - width / 2,
    y: needClipY ? top : top - height / 2,
    width: needClipX ? width : width * 2,
    height: needClipY ? height : height * 2
  }));
}
